nucypher/blockchain/eth/signers/base.py

```diff
@@ -0,0 +1,94 @@
+"""
+ This file is part of nucypher.
+
+ nucypher is free software: you can redistribute it and/or modify
+ it under the terms of the GNU Affero General Public License as published by
+ the Free Software Foundation, either version 3 of the License, or
+ (at your option) any later version.
+
+ nucypher is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU Affero General Public License for more details.
+
+ You should have received a copy of the GNU Affero General Public License
+ along with nucypher.  If not, see <https://www.gnu.org/licenses/>.
+"""
+
+from abc import ABC, abstractmethod
+from typing import List
+from urllib.parse import urlparse
+
+from hexbytes.main import HexBytes
+
+from nucypher.utilities.logging import Logger
+
+
+class Signer(ABC):
+
+    URI_SCHEME = NotImplemented
```

Comment:
I rather like the overriding pattern of - Is there an advantage of moving it to a method?

Reply:
The same advantage you get from all the other methods marked as abstract - its existence will be enforced. It just seems strange to me to have direct redefinition mixed with `ABC` usage in the same class.

Of course, the problem here is that it's a static method, and `abc` can only enforce definition of abstract methods on class instantiation (and it doesn't check that the defined method is actually static). Still, it would provide some protection from mistakes.

Reply:
> The same advantage you get from all the other methods marked as abstract - its existence will be enforced.

That, to me, is a great motivation to do it.

Reply:
While I understand the intention here I cannot see the implementation:  In order to make this a property, it needs to be an instance method, however, I actually need a class-scoped attribute for use before instantiation.

How shall I proceed?

Reply:
Additionally - in terms of preventing human error - leaving it as a `NotImplemented` class attribute will still result in a loud and early failure.

Reply:
> In order to make this a property, it needs to be an instance method, however, I actually need a class-scoped attribute for use before instantiation.

You're right, you can't enforce a "class property" in Python, unfortunately. I personally would choose restriction over syntactic sugar here and just use a static classmethod, but I do not insist on it.

> in terms of preventing human error - leaving it as a NotImplemented class attribute will still result in a loud and early failure.

My point was that all the other attributes of this base class already use `abc` machinery, and only this specific one relies on `NotImplemented` - so it seems that we agree that `abc` is a superior approach.

Reply:
Pushed an update as an abstract classmethod - let me know what you think.

Reply:
Looks great to me.